src/System.Management.Automation/engine/CommandCompletion/CompletionHelpers.cs

```diff
@@ -21,36 +21,67 @@ internal static class CompletionHelpers
 
             foreach (string value in possibleCompletionValues)
             {
-                if (pattern.IsMatch(value))
+                if (IsMatch(value, wordToComplete))
                 {
                     string completionText = QuoteCompletionText(value, quote);
+                    string toolTip = toolTipMapping?.Invoke(value) ?? value;
+                    string listItemText = listItemTextMapping?.Invoke(value) ?? value;
 
-                    string listItemText = value;
-
-                    yield return new CompletionResult(
-                        completionText,
-                        listItemText,
-                        resultType,
-                        toolTip: toolTipMapping is null
-                            ? listItemText
-                            : toolTipMapping(value));
+                    yield return new CompletionResult(completionText, listItemText, resultType, toolTip);
                 }
             }
         }
 
+        /// <summary>
+        /// Determines whether the given value matches the specified word or pattern.
+        /// </summary>
+        /// <param name="value">The input string to check for a match.</param>
+        /// <param name="wordToComplete">The word or partial word to compare against the input string.</param>
+        /// <returns>
+        /// Returns <c>true</c> if the value matches the wordToComplete or the generated wildcard pattern; otherwise, <c>false</c>.
+        /// </returns>
+        /// <remarks>
+        /// The method performs the following checks:
+        /// 1. If the value contains escaped newline characters, the wordToComplete is normalized 
+        ///    and unescaped, and a case-insensitive prefix match is performed.
+        /// 2. If either the value or wordToComplete contains wildcard characters, a case-insensitive 
+        ///    prefix match is performed. This is to protect against issues in WildcardPatternParser.Parse()
+        ///    where strings like '[*' throw WildcardPatternException from wildcards not being escaped.
+        /// 3. If neither of the above conditions apply, a wildcard pattern is generated from the 
+        ///    wordToComplete, appending a wildcard character (*). The pattern is then used to match the value.
+        /// </remarks>
+        internal static bool IsMatch(string value, string wordToComplete)
+        {
+            if (ContainsEscapedNewlineString(value))
+            {
+                string normalizedWord = WildcardPattern.Unescape(wordToComplete.ReplaceLineEndings("`"));
+                return value.StartsWith(normalizedWord, StringComparison.OrdinalIgnoreCase);
+            }
+            else if (WildcardPattern.ContainsWildcardCharacters(value) ||
+                     WildcardPattern.ContainsWildcardCharacters(wordToComplete))
+            {
+                return value.StartsWith(wordToComplete, StringComparison.OrdinalIgnoreCase);
+            }
```

Comment:
Do we really need to evaluate these both conditions?
```c#
string normalizedWord = WildcardPattern.Unescape(wordToComplete.ReplaceLineEndings("`"));
if (!value.StartsWith(normalizedWord, StringComparison.OrdinalIgnoreCase))
{
    return WildcardPattern
                .Get(wordToComplete + "*", WildcardOptions.IgnoreCase)
                .IsMatch(value);
}
```

Reply:
My thoughts are that `Unescape()` and `ReplaceLineEndings()` returns source string if there is nothing to do. So, no need to do checks before running the methods. Then if literal comparison returns true the method returns true. Otherwise, we do wildcard comparison.

```c#
string normalizedWord = WildcardPattern.Unescape(wordToComplete.ReplaceLineEndings("`"));
if (!value.StartsWith(normalizedWord, StringComparison.OrdinalIgnoreCase))
{
    return value.StartsWith(normalizedWord, StringComparison.OrdinalIgnoreCase))
        ? true
        : WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase) .IsMatch(value);
}
```

Reply:
I think with this one it gets tricky if `value` is `word` and `wordToComplete includes wildcards like `word*`. If we use wildcard pattern `IsMatch` it will return `true`, but `StartsWith` will return `false`.

I guess first one makes more sense in terms of behavior?

Reply:
In that case maybe something like this will be clearer:

```csharp
internal static bool IsMatch(string value, string wordToComplete)
        {
            string normalizedWord = WildcardPattern.Unescape(wordToComplete.ReplaceLineEndings("`"));

            bool match = value.StartsWith(normalizedWord, StringComparison.OrdinalIgnoreCase);

            if (!match)
            {
                match = WildcardPattern
                    .Get(wordToComplete + "*", WildcardOptions.IgnoreCase)
                    .IsMatch(value);
            }

            return match;
        }
```

Reply:
LGTM.

> it gets tricky

We can always add a comment to explain how the method works.